### PR TITLE
Division in a finite extension

### PR DESCRIPTION
**Pull request by jksuom**
#### Brief description of what is fixed or changed

Define division in a finite extension.
#### Release Notes

* polys
    * Add division to finite extensions.

### CODE REVIEW

**Comment by gschintgen** [comment]
I checked this out locally and did a few simple computations in GF(27) and QQ[x]/(x**2-2). Nice work.
So I suppose this only needs a few tests? (And ideally a minimum of public documentation.)

I see the commit is actually from 2018. Do you intend to follow up on this?

**Comment by jksuom** [comment]
> this only needs a few tests

That is why I didn't push it at the time. There seemed to be some interest in improving the FF class, and I was planning to add this as a part of it. But that interest faded away (and I also seemed to have more urgent tasks :).

You are welcome to add tests and documentation.

**Comment by gschintgen** [comment]
> You are welcome to add tests and documentation.

Ok, will do. Thanks.

**Comment by gschintgen** [comment]
Tests have failed due to timeout:
```
sympy/solvers/ode/tests/test_single.py[17] ..www.........

The job exceeded the maximum time limit for jobs, and has been terminated.
```
https://travis-ci.org/github/sympy/sympy/jobs/700338224#L528

I'll add at least one more commit anyway, so let's see how that one goes. And hopefully Github picks up on the results...

**Comment by oscarbenjamin** [comment]
> Tests have failed due to timeout:

That's because of #19536

**Comment by gschintgen** [comment]
I noticed that a nice exposition had already been contributed by @jksuom. Please review.
Here's how the docs will look:

Screenshot_2020-06-21 AGCA - Algebraic Geometry and Commutative Algebra Module — SymPy 1 7 dev documentation(4)

File: sympy/polys/agca/extensions.py [hunk 1]
```diff
@@ -109,12 +153,51 @@ def __str__(f):
...
+
+    >>> t = Symbol('t')
+    >>> FiniteExtension(Poly(t**2 - x**3 - x + 1, t, field=True))
+    ZZ(x)[t]/(t**2 - x**3 - x + 1)
+
+    Notes
+    =====
+
+    It's not possible to use a ``FiniteExtension`` as ``domain`` for the
+    :class:`~.Poly` class.
```
Comment (oscarbenjamin) [inline sympy/polys/agca/extensions.py:199]: Why can't this be used as a domain?

Comment (jksuom) [reply sympy/polys/agca/extensions.py:199]: A more a fitting comment might be that `FiniteExtension` is not a subclass of `Domain`. However, it should be possible to define a `FiniteField` class that would derive from `Domain` and `FiniteExtension`. It is also possible to define other wrapper subclasses of `Domain` for (some types of)  `FiniteExtension`. (Consider `PolynomialRing` vs. `PolyRing`.)

Comment (gschintgen) [reply sympy/polys/agca/extensions.py:199]: > A more a fitting comment might be that `FiniteExtension` is not a subclass of `Domain`. However, it should be possible to define a `FiniteField` class that would derive from `Domain` and `FiniteExtension`. It is also possible to define other wrapper subclasses of `Domain` for (some types of) `FiniteExtension`. (Consider `PolynomialRing` vs. `PolyRing`.)

I'll see what this entails and how this works out and at the very least improve the comment.

Finite fields were the motivation for looking into this, but in the meantime I've also had a closer look at `galoistools.py` and it definitely has everything that would be needed for implementing proper non-prime finite fields (including algorithms for determining irreducibility over Z_p and actually finding irreducible polynomials). So maybe that's the better way forward when it comes to implementing GF(p^n) as polys domain. At least it seems to have been the original plan.

Comment (gschintgen) [reply sympy/polys/agca/extensions.py:199]: I'm trying to work out a proof-of-concept for GF(p**d) using FiniteExtension, but I must confess that I'm struggling a bit with all the (messy/undocumented/labyrinthine) internals of `polys`. In particular:
- what is the `lev` argument of `DMP`?
https://github.com/sympy/sympy/blob/a69a68cf278728663c9ced8a9b1a1cb4dfc6355a/sympy/polys/polyclasses.py#L145-L162
- I noticed the following:
```python
In [60]: F = FiniteExtension(poly(x**3+x**2+2, modulus=5))

In [61]: type(F.ring)
Out[61]: sympy.polys.domains.old_polynomialring.GlobalPolynomialRing
```
Isn't the `old_polynomialring` somewhat deprecated? Does it matter for this application?

Comment (oscarbenjamin) [reply sympy/polys/agca/extensions.py:199]: I think the lev argument is basically the number of generators:
```julia
In [5]: Poly(x, x, domain=ZZ).rep.lev                                                                                                          
Out[5]: 0

In [6]: Poly(x, x, y, domain=ZZ).rep.lev                                                                                                       
Out[6]: 1

In [7]: Poly(x, x, y, z, domain=ZZ).rep.lev                                                                                                    
Out[7]: 2
```
I think it's called `lev` as short for "level" representing the fact that in the dense representation a poly in `n` generators is implemented as a poly of polys in `n-1` generators. The internal representation is a list of domain elements for an univariate poly and then a list of lists for bivariate etc:
```julia
In [9]: Poly(x, x, domain=ZZ).rep.rep                                                                                                          
Out[9]: [mpz(1), mpz(0)]

In [10]: Poly(x, x, y, domain=ZZ).rep.rep                                                                                                      
Out[10]: [[mpz(1)], []]

In [11]: Poly(x, x, y, z, domain=ZZ).rep.rep                                                                                                   
Out[11]: [[[mpz(1)]], [[]]]
```
The `lev` argument basically tells the `dmp_*` functions how many levels to recurse.

Comment (gschintgen) [reply sympy/polys/agca/extensions.py:199]: > I think the lev argument is basically the number of generators [...]

Thanks!

I'm also hitting circular import issues (concerning `Poly`) when I try adding a `PrimePowerFiniteField` class in `domains/finitefield.py` that subclasses `FiniteField` and `FiniteExtension`. That's probaby not unexpected given that logically the chain is domains -> polynomials -> extensions. Maybe a more direct implementation is less prone to such issues.

Concerning this PR I'll modify the note in the docstring and that's it.

Comment (jksuom) [reply sympy/polys/agca/extensions.py:200]: > Isn't the old_polynomialring somewhat deprecated?

It depends on the number of generators. For a small number of generators, say, 1, 2, (3,) and dense polynomials, I think that it is quite efficient. For a large number of generators, the nested list representation becomes very inefficient.

Comment (jksuom) [reply sympy/polys/agca/extensions.py:200]: `lev` is the number of generators minus 1.

Comment (jksuom) [reply sympy/polys/agca/extensions.py:200]: > 'm also hitting circular import issues (concerning Poly) when I try adding a PrimePowerFiniteField class in domains/finitefield.py that subclasses FiniteField and FiniteExtension.

I have suggested another approach: Define a new class, say, `FiniteRing`, that would do what `FiniteField` does now (or simply rename `FiniteField`). Invoking `FiniteField` with a non-prime argument would pass the call to `FiniteRing`, presumably with a deprecation statement. Then `FiniteField` could subclass from `FiniteExtension`.

**Comment by gschintgen** [comment]
The Travis build hit the timeout again... :-/

File: sympy/polys/agca/extensions.py (hunk 1 above)
Comment (gschintgen) [reply sympy/polys/agca/extensions.py:200]: > presumably with a deprecation statement

I don't think that's necessary as long as the new `FiniteField` also accepts the current parameters for creating a finite field of prime order. There are only some 3 or 4 very minor (and non-sensical) tests in the whole testsuite that actually instantiate a ring of integers modulo a composite. There's no real use anywhere and the current broken behavior is (fortunately) not publicly documented.

I think it's best to take the discussion concerning the implementation to #9544.

**Comment by oscarbenjamin** [comment]
I restarted the timed out tests which have now finished. Unfortunately there are no merge conflicts...

**Comment by gschintgen** [comment]
> 
> 
> I restarted the timed out tests which have now finished. Unfortunately there are no merge conflicts...

Strange, locally I could now merge master to the branch without a hitch. Hmm. I'll push that merge commit. That should do.

**Comment by gschintgen** [comment]
And now the timeout hit again. I think this can wait till the slow tests/density split have been sorted out. I'll refrain from restarting the job to leave more Travis capacity for those more urgent issues...

**Comment by gschintgen** [comment]
Tests have now passed successfully.

**Comment by oscarbenjamin** [comment]
Looks good

**Comment by oscarbenjamin** [comment]
Thanks @jksuom and @gschintgen!

**Pull request merged**